src/lz4.rs

```diff
@@ -524,19 +531,24 @@ impl<W: Writer> Encoder<W> {
     /// This function is used to flag that this session of compression is done
     /// with. The stream is finished up (final bytes are written), and then the
     /// wrapped writer is returned.
-    pub fn finish(mut self) -> (W, old_io::IoResult<()>) {
-        let result = self.flush();
-        let result = result.and(self.w.write_le_u32(0));
-        // XXX: this checksum is wrong
-        let result = result.and(self.w.write_le_u32(0));
+    pub fn finish(mut self) -> (W, io::Result<()>) {
+        let mut result = self.flush();
+
+        for _ in 0..2 {
+            let tmp = self.w.write_u32::<LittleEndian>(0)
```

Comment:
This is the only point I didn't address. I don't see how a fold would work here.. Could you give me a pointer?